test/fixtures/apps/gzip-fetch.js

```diff
@@ -0,0 +1,6 @@
+fly.http.respondWith(async function(req){
+  let resp = await fetch("https://example.com", { headers: {"Accept-Encoding": "gzip"}, inflate: true })
+  let txt = await resp.text()
+  console.log("Got text: ", txt.length)
+  return new Response(txt, resp)
+})
```

Comment:
To make this work end-to-end, you'll probably want to use a different test URL. Maybe add query parameters here (example: https://example.com/?gzip). Then run your test with the `REPLAY=record` environment variable. This will tell node-replay to do the request and save the response in our fixtures.

Reply:
^ We don't allow external requests to go out during tests because it's unpredictable. Our tests would be slow and might fail if the endpoint is not available or fails entirely. Therefore we record and store all HTTP requests and responses we need, automatically, with node-replay. You can see them in `test/fixtures/replay/`.

Anyway, by default no external requests are allowed.

In the case of this test, it might request this fixture: https://github.com/superfly/fly/blob/master/test/fixtures/replay/example.com-443/basic-fetch depending how it matches the request.

If it fails with a network error, then it means it's trying to match a request we don't have recorded and stored. That's good. That's when you want to use the `REPLAY=record` environment variable.

Also, I think @mrkurt mentioned node-replay automatically inflated gzipped responses. That seems bad. We had to add our own changes to node-replay too and that's why it's vendored in `vendor/node-replay/`. This will require some digging.

Reply:
When I tried the test again it seemed to work - but it did not seem to save anything or decode the data (probably because like you mentioned node-replay is doing that). I will have a look a node-replay in a minute. 

I looked in `test/fixtures/replay/` and nothing new was saved. I exported `REPLAY=record`, but it seemed to run the tests in the same way - am I missing something?

Thanks for the feedback.

Reply:
Never mind, I figured it out. I had not set up my test in `server.spect.ts`.

Reply:
You where right - now that I am getting the response, it is unzipping it. But I am also getting an error. 
```
  1) Server
       basic gzip fetch and read app
         may fetch responses externally:
     Error: Parse Error
      at Socket.socketOnData (_http_client.js:440:20)
      at addChunk (_stream_readable.js:263:12)
      at readableAddChunk (_stream_readable.js:250:11)
      at Socket.Readable.push (_stream_readable.js:208:10)
      at TCP.onread (net.js:607:20)
```
I think this might have something to do with the stream not knowing how much data there is because the content-length might be telling it to stop getting the data early. 

Source: https://stackoverflow.com/questions/10838186/nodejs-error-parse-error-at-socket-ondata

Reply:
Changing the `content-length` in replay did seem to work. But it is still only text not gzip data.

Reply:
Oh yeah, that's interesting. If replay is inflating and not changing the content-length header that would cause all kinds of problems.

Reply:
Yeah, my guess is that is what happened.

Reply:
I think I got rid of the problem with vendor. I just commented out where it decoded the gzip data - its not a great solution but, it will work for testing. Now it does not seem to be able to inflate it.

Reply:
Just updated my code.

Back to the original error now 😞 
```
     Error: Timeout of 5000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.
```
Thanks again for being patient and helping explain this to me.

Reply:
Hmmm .. that error means something isn't finishing.

It'll probably help to add logging to `src/bridge/proxy_stream.ts`. That _should_ work with any kind of readable. It'll help debug if you can actually see data hitting somewhere like here: https://github.com/superfly/fly/blob/master/src/bridge/proxy_stream.ts#L97

Also I think I just realized what was happening with those headers. Instead of changing the replay source to remove them, you should put that code in the `fetch` class. When you do the gunzip logic you should also `res.headers.delete("content-length")` and `res.headers.delete("content-encoding")`

Reply:
It actually looks like it is not getting any data. That is probably the problem. I will look into that in a second and see if I can figure it out.

Reply:
My guess is that the issue is here:
```JavaScript
zipStream = res.pipe(zlib.createUnzip())
```
Is there any way to get the body from the response? I think if we could parse the body instead of using a stream from the response that might work better.

Reply:
We have to use streams because people can `fetch` very large files which then gets proxied to an end visitor.  Trying to load all that in memory at once would be brutal.

Reply:
I _think_ you're on the right track. You could try swapping out that proxy stream for `process.stdout` to see if it's doing anything at all.

Reply:
I tried that earlier and it did not output anything. I think the issue is I did not set `trailers`. Because in `replay/lib/pass_through.js` it has something like this:
```JavaScript
        const unzip = zlib.createUnzip();
        unzip.trailers = res.trailers;
        unzip.rawTrailers = res.rawTrailers;
        zipStream = res.pipe(unzip);
```
But, typescript is complaining about that so, I will try to keep playing around with it.

Reply:
I just had a thought, are we trying to inflate the `res` or the `req`? It would make sense to me that we would be trying to inflate the request.

Reply:
Mostly the `res` (the response we get from the other server). People write code to make requests, then change the response body as it comes through.

Reply:
Oh, that makes sense now. I feel like when I look at the same code for a long time, sometimes I think stuff that makes no sense 😆